### PR TITLE
Bug 1612286 - Single sign-on through Mozilla IAM

### PR DESCRIPTION
**Pull request by dklawren**

### CODE REVIEW

File: oauth2.cgi [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# Get access token from OAuth2 provider;
+my $resp = $c->oauth2->get_token('oauth2');
+
+# Store access token for use by OAuth2 login info getter
+$cache->{oauth2_client_userinfo} = $c->oauth2->userinfo($resp->{access_token});
+
+my $user = Bugzilla->login(LOGIN_REQUIRED);
+
+if ($user && $user->in_group('no-sso-auth')) {
```
Comment (globau) [inline oauth2.cgi:34]: Can you give examples of why we'd want to exclude users that are "too powerful" – that seems to be the opposite to what we should be doing.

File: oauth2.cgi [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env perl
```
Comment (globau) [inline oauth2.cgi:1]: shouldn't this be a mojo route rather than a cgi?

File: template/en/default/account/auth/login.html.tmpl [hunk 3]
```diff
@@ -84,6 +84,12 @@
   </form>
 </div>
 
+[% IF Param('oauth2_client_enabled') %]
+  [% INCLUDE account/auth/oauth2_login_button.html.tmpl
+     target => target
+  %]
+[% END %]
```
Comment (globau) [inline template/en/default/account/auth/login.html.tmpl:91]: Can you please include a screenshot of how the two login buttons are presented

**Review by globau — CHANGES_REQUESTED** [review]
I see this is WIP so some quick comments.

File: oauth2.cgi (hunk 1 above)
Comment (dklawren) [reply oauth2.cgi:34]: I was duplicating what we do currently with Github auth where we do not allow people with heightened security permissions to login via Github. But thinking it about it more now, it is OK to allow users of Mozilla IAM to be able to login regardless of permissions. And the fact that we could one day make all people use SSO for auth. Will remove.

File: oauth2.cgi (hunk 2 above)
Comment (dklawren) [reply oauth2.cgi:1]: That was my original plan. But after trying various hacks to make it work I gave in and made it a small cgi. Reason is that the way our current Auth stack works relies heavily on running inside the CGI compatibility wrapper that we use with Mojolicious. I would either need to do quite a bit of refactoring or duplicating a lot of code to make it work. 

My future plan is to start moving most of the supporting code (login, utilities, etc.) into native Mojo and then doing these things will be much less painful.

File: template/en/default/account/auth/login.html.tmpl (hunk 3 above)
Comment (dklawren) [reply template/en/default/account/auth/login.html.tmpl:91]: image

image

Comment (globau) [reply template/en/default/account/auth/login.html.tmpl:91]: It'll need to be styled to match GitHub's button (or vice versa, although having the service's icon is super useful).

Comment (dklawren) [reply template/en/default/account/auth/login.html.tmpl:91]: Probably should use something more standard such as 

https://lipis.github.io/bootstrap-social/

This is what Lando has currently

**Pull request closed without merging**